src/VarArray.mo

```diff
@@ -107,104 +223,740 @@ module {
+  /// and concatenating the resulting arrays in order.
+  ///
+  /// ```motoko include=import
+  /// import Nat "mo:base/Nat";
+  ///
+  /// let array = [var 1, 2, 3, 4];
+  /// VarArray.flatMap<Nat, Int>(array, func x = [x, -x])
+  ///
+  /// ```
+  /// Runtime: O(size)
+  ///
+  /// Space: O(size)
+  /// *Runtime and space assumes that `k` runs in O(1) time and space.
+  public func flatMap<T, R>(array : [var T], k : T -> Iter.Iter<R>) : [var R] {
+    var flatSize = 0;
+    let arrays = Prim.Array_tabulate<[var R]>(
+      array.size(),
+      func i {
+        let subArray = fromIter<R>(k(array[i])); // TODO: optimize
+        flatSize += subArray.size();
+        subArray
+      }
+    );
+
+    // could replace with a call to flatten,
+    // but it would require an extra pass (to compute `flatSize`)
+    var outer = 0;
+    var inner = 0;
+    tabulate<R>(
+      flatSize,
+      func _ {
+        while (inner == arrays[outer].size()) {
+          inner := 0;
+          outer += 1
+        };
+        let element = arrays[outer][inner];
+        inner += 1;
+        element
+      }
+    )
   };
 
+  /// Collapses the elements in `array` into a single value by starting with `base`
+  /// and progessively combining elements into `base` with `combine`. Iteration runs
+  /// left to right.
+  ///
+  /// ```motoko include=import
+  /// import {add} "mo:base/Nat";
+  ///
+  /// let array = [var 4, 2, 0, 1];
+  /// let sum =
+  ///   VarArray.foldLeft<Nat, Nat>(
+  ///     array,
+  ///     0, // start the sum at 0
+  ///     func(sumSoFar, x) = sumSoFar + x // this entire function can be replaced with `add`!
+  ///   );
+  /// ```
+  ///
+  /// Runtime: O(size)
+  ///
+  /// Space: O(1)
+  ///
+  /// *Runtime and space assumes that `combine` runs in O(1) time and space.
   public func foldLeft<T, A>(array : [var T], base : A, combine : (A, T) -> A) : A {
-    todo()
+    var acc = base;
+    for (element in array.vals()) {
+      acc := combine(acc, element)
+    };
+    acc
   };
 
+  /// Collapses the elements in `array` into a single value by starting with `base`
+  /// and progessively combining elements into `base` with `combine`. Iteration runs
+  /// right to left.
+  ///
+  /// ```motoko include=import
+  /// import {toText} "mo:base/Nat";
+  ///
+  /// let array = [1, 9, 4, 8];
+  /// let bookTitle = VarArray.foldRight<Nat, Text>(array, "", func(x, acc) = toText(x) # acc);
+  /// ```
+  ///
+  /// Runtime: O(size)
+  ///
+  /// Space: O(1)
+  ///
+  /// *Runtime and space assumes that `combine` runs in O(1) time and space.
   public func foldRight<T, A>(array : [var T], base : A, combine : (T, A) -> A) : A {
-    todo()
+    var acc = base;
+    let size = array.size();
+    var i = size;
+    while (i > 0) {
+      i -= 1;
+      acc := combine(array[i], acc)
+    };
+    acc
   };
 
+  /// Flattens the iterator of mutable arrays into a single mutable array. Retains the original
```

Comment:
Again, I thin it would be good to keep the nested array version and then define one for iterators on top of that if possible.